### PR TITLE
Major refactoring required to work with actual proxy

### PR DESCRIPTION
**Pull request by tufandevrim**
## WHAT
To utilize main browserify's enties, plugins, options, etc.. thru actual proxy method in lazy browserifies, this major refactoring required. This is especially necessary to use the plugins registered in main bundle in lazy bundles.

## HOW
- Refactor main flow and separate action function into extractify prototype
- Remove extractify plugin in lazy bundles to prevent looping
- updated eslint rules
- Extra making external steps in lazyBundle

### CODE REVIEW

**Comment by yahoocla** [comment]
Thank you for submitting this pull request, however I do not see a valid CLA on file for you. Before we can merge this request please visit https://yahoocla.herokuapp.com/ and agree to the terms. Thanks! :smile:

**Comment by yahoocla** [comment]
CLA is valid!

File: index.js [hunk 1]
```diff
@@ -9,229 +9,97 @@ var xtend = require('xtend');
...
+
+    self.opts = processOptions(opts);
+    self.mainExternals = [];
+    self.moduleBundleMap = {};
+    self.lazyBundleMapDestination = self.opts.bundleMapOption ? self.opts.bundleMapOption.dest : false;
+    self.lazyBundleMapBaseDir = self.opts.bundleMapOption ? self.opts.bundleMapOption.basedir : false;
+    self.lazyBundleMapInjectSoft = true;
+
+    if (self.opts.bundleMapOption &&
+        (self.opts.bundleMapOption.injectSoft === false || self.opts.bundleMapOption.injectSoft === 'false')) {
```
Comment (juandopazo) [inline index.js]: `=== 'false'`? 👽

File: index.js [hunk 2]
```diff
@@ -407,4 +213,241 @@ module.exports = function extractify(b, opts) {
...
+                fs.writeFileSync(path.resolve(self.basedir, self.lazyBundleMapDestination),
+                    JSON.stringify(self.moduleBundleMap, null, 4), {encoding: 'utf8'});
+                injectOnce = false;
+            }
+        }
+
+        this.push(row);
+        next();
+    }));
+}
```
Comment (juandopazo) [inline index.js]: Oh, the "sometimes" semicolons... 😜

**Comment by juandopazo** [comment]
It's hard to review all the actual changes, but overall it looks like a step in the right direction.

**Comment by irae** [comment]
I like that it is clear what are pure functions, what are instance methods and the structure is way better. Most importantly it is pretty clear that the proxy is being called in sync, which should solve the main problems.

👍 

I would say it is a good idea to add tests and examples, either as part of this PR or soon after. The fact that this didn't break previous tests is already good. But something similar to the example we discussed before (the programmatic use of bundle collapser) would be a nice addition to the tests.

**Comment by gomezd** [comment]
👍

File: index.js (hunk 1 above)
Comment (tufandevrim) [inline index.js]: Yeah it comes as `false` from the command line.

**Pull request merged**